packages/test/test-drivers/src/factory.ts

```diff
@@ -28,26 +26,35 @@ export const DriverApi: DriverApiType = {
     RouterliciousDriverApi,
 };
 
-let httpAgent: http.Agent | undefined;
-function setKeepAlive(api: RouterliciousDriverApiType) {
-    // Each TCP connect has a delay to disallow it to be reused after close, and unit test make a lot of connection,
-    // which might cause port exhaustion.
+let httpRequestPatched = false;
+function patchHttpRequestToForceKeepAlive() {
+    // Each TCP connection port has a delay to disallow it to be reused after close,
+    // and unit test make a lot of connection, which might cause port exhaustion.
+    // Patch http.request to force keep-alive.
 
-    // For drivers that use Axios (t9s and r11s), keep the TCP connection open so that they can be reused
-    // TODO: no solution for node-fetch used by ODSP driver.
-    // TODO: currently the driver use a global setting.  Might want to make this encapsulated.
+    if (httpRequestPatched) { return; }
 
-    // create the keepAlive httpAgent only once
-    if (httpAgent === undefined) {
-        httpAgent = new http.Agent({ keepAlive: true });
-    }
+    httpRequestPatched = true;
 
-    // eslint-disable-next-line @typescript-eslint/no-require-imports
-    const axios = api.modulePath === "" ? Axios : require(path.join(api.modulePath, "node_modules", "axios"));
-    // Don't override it if there is already one
-    if (axios.defaults.httpAgent === undefined) {
-        axios.defaults.httpAgent = httpAgent;
-    }
+    const httpAgent = new http.Agent({ keepAlive: true, maxFreeSockets: Infinity });
+    const oldRequest = http.request;
+    http.request = ((url, options, callback) => {
+        let opts;
+        if (options !== undefined) {
+            opts = typeof options !== "function" ? options : url;
+        } else if (callback !== undefined) {
+            // eslint-disable-next-line no-param-reassign
+            options = {};
+            opts = options;
+        } else {
+            opts = url;
+        }
+        if (opts.agent === undefined) {
+            opts.agent = httpAgent;
+            opts.headers.Connection = ["keep-alive"];
+        }
+        return oldRequest(url, options, callback);
```

Comment:
I'm not quite getting how this works... Shouldn't this line pass in `opts` instead of `options`? Otherwise I don't understand what's the point of the `opts` object... but there's a chance that `opts` is just a string (url) with `agents` and `headers` properties, which I'm not sure would make sense as second argument here...

Reply:
opts the variable that determines where we should put the httpAgent.  But the argument positionally stays the same.

e.g. if the first argument is the options object, we will opts === url, and we will put it there.  but we would still pass that as the first argument.

Reply:
Is url an object then?

Reply:
it could be a string or it could be an object.  The function params are overloaded.  That's why all the "if" logic there to determine which param is the options object. 

The two variants are:

- `http.request(options[, callback])`
- `http.request(url[, options][, callback])`

See https://nodejs.org/dist/latest-v18.x/docs/api/http.html#httprequestoptions-callback

Reply:
> The two variants are:

> http.request(options[, callback])
> http.request(url[, options][, callback])
> See https://nodejs.org/dist/latest-v18.x/docs/api/http.html#httprequestoptions-callback

nit - Adding these lines + link to the comment would be wonderful

Reply:
Added.